File: 002.Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head, *dummy, *result;
        head = dummy = result = new ListNode(0);
        int carry = 0;
        if(l1 == NULL || l2 == NULL)
            return result;
            
        while(l1 != NULL || l2 != NULL) {
            int sum = 0;
            if (l1 != NULL && l2 != NULL) {
                sum = l1->val + l2->val + carry;
                l1 = l1->next;
                l2 = l2->next;
            }
            else if(l1 != NULL) {
                sum = l1->val + carry;
                l1 = l1->next;
            }
            else {
                sum = l2->val + carry;
                l2 = l2->next;
            }
            carry = sum / 10;
            result->next = new ListNode(sum % 10);
            result = result->next;
        }
        
        if(carry != 0)
            result->next = new ListNode(carry);
        
        head = dummy->next;
        delete dummy;
        
        return head;
    }
};

File: 002. Add Two Numbers.java
//    You are given two non-empty linked lists representing two non-negative integers.
//    The digits are stored in reverse order and each of their nodes contain a single digit.
//    Add the two numbers and return it as a linked list.
//
//    You may assume the two numbers do not contain any leading zero, except the number 0 itself.
//
//    Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
//    Output: 7 -> 0 -> 8

    /**
     * Definition for singly-linked list.
     * public class ListNode {
     *     int val;
     *     ListNode next;
     *     ListNode(int x) { val = x; }
     * }
     */

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode(0);
        ListNode next = head;
        int carry = 0;

        while (l1 != null || l2 != null || carry != 0) {
            int result = ((l1 != null) ?l1.val : 0) + ((l2 != null) ?l2.val : 0) + carry;
            carry = result / 10;
            next.next = new ListNode(result % 10);
            next = next.next;
            l1 = (l1 != null) ? l1.next : l1;
            l2 = (l2 != null) ? l2.next : l2;
        }

        return head.next;
    }
